### PR TITLE
Scale Upgrade testcase 

### PR DESCRIPTION
**Pull request by ramkiperiy**
Scale pre_upgrade and post_upgrade simple testcase
TC Scales 1500 PVCs, expected OSDs min 3

Signed-off-by: Ramakrishnan <rperiyas@redhat.com>

### CODE REVIEW

File: tests/ecosystem/upgrade/test_upgrade_with_scaled_pvcs_pods.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        pvc_data = all_pvc_dict["items"][i]
+        if not pvc_data["status"]["phase"] == constants.STATUS_BOUND:
+            pvc_not_bound_list.append(pvc_data["metadata"]["name"])
+        else:
+            pvc_bound_list.append(pvc_data["metadata"]["name"])
+
+    # Get all PODs from namespace
+    ocp_pod_obj = OCP(kind=constants.DEPLOYMENTCONFIG, namespace=namespace)
+    all_pods_dict = ocp_pod_obj.get()
+    pod_running_list, pod_not_running_list = ([] for i in range(2))
```
[the remark this answers is not in the document]
Comment (ramkiperiy) [reply tests/ecosystem/upgrade/test_upgrade_with_scaled_pvcs_pods.py:114]: I didn't know about this, it's cool !!!. Looks like we can declare more lists a, b, c = ([], [], [])

[the remark this answers is not in the document]
Comment (ramkiperiy) [reply tests/ecosystem/upgrade/test_upgrade_with_scaled_pvcs_pods.py:114]: Changed the code like below, 
```
    (pvc_bound_list, pvc_not_bound_list, pod_running_list, pod_not_running_list) = (
        [], [], [], []
    )
```
black is reformatting  above line 
`    (pvc_bound_list, pvc_not_bound_list, pod_running_list, pod_not_running_list) = (
        [],
        [],
        [],
        []
    )`
We can also use something like below
`    (pvc_bound_list, pvc_not_bound_list, pod_running_list, pod_not_running_list) = (
        [] for i in range(4)
    )`
for now changing the code with above declaration but anyway let me know your suggestion.

[the remark this answers is not in the document]
Comment (ramkiperiy) [reply tests/ecosystem/upgrade/test_upgrade_with_scaled_pvcs_pods.py]: @vasukulkarni Updated the code, still declaring it like below will reduce 2 lines of code, let me know
```
    (pvc_bound_list, pvc_not_bound_list, pod_running_list, pod_not_running_list) = (
        [] for i in range(4)
    )
```

**Comment by ramkiperiy** [comment]
> PR validation
> 
> Cluster Name:
> Cluster Configuration:
> PR Test Suite: pre_ocs_upgrade or ocs_upgrade or post_ocs_upgrade
> PR Test Path: tests/ecosystem/upgrade/test_upgrade_with_scaled_pvcs_pods.py
> Additional Test Params:
> OCP VERSION: 4.7
> OCS VERSION: 4.6
> 
> [Job UNSTABLE](https://ocs4-jenkins-csb-ocsqe.apps.ocp4.prod.psi.redhat.com/job/qe-trigger-test-pr/629//testReport) (some or all tests failed).

Failed jenkins run analysis --> **Setup deployed with OCS 4.6(passed), executed pre_ocs_upgrade test to create 1500 PVCs and PODs (passed), Upgrade to OCS 4.7(passed), pod_ocs_upgrade test to validate PODs and PVCs count (failed) due to issue in for loop range(value)** --> have fixed this problem expecting the test case to pass in next validation run

**Review by vasukulkarni — DISMISSED** [review]

**Comment by ramkiperiy** [comment]
> PR validation
> 
> Cluster Name:
> Cluster Configuration:
> PR Test Suite: pre_ocs_upgrade or ocs_upgrade or post_ocs_upgrade
> PR Test Path: tests/ecosystem/upgrade/test_upgrade_with_scaled_pvcs_pods.py
> Additional Test Params:
> OCP VERSION: 4.7
> OCS VERSION: 4.6
> 
> [Job UNSTABLE](https://ocs4-jenkins-csb-ocsqe.apps.ocp4.prod.psi.redhat.com/job/qe-trigger-test-pr/636//testReport) (some or all tests failed).

Validation failed due to one of the DC pod was in CrashLoop state, this is intermittent problem have to add exceptions to catch such problems so that those pods can be deleted or bring it back to running state, will take this in upcoming PRs

**Comment by ramkiperiy** [comment]
> PR validation
> 
> Cluster Name:
> Cluster Configuration:
> PR Test Suite: pre_ocs_upgrade or ocs_upgrade or post_ocs_upgrade
> PR Test Path: tests/ecosystem/upgrade/test_upgrade_with_scaled_pvcs_pods.py
> Additional Test Params:
> OCP VERSION: 4.7
> OCS VERSION: 4.6
> 
> [Job UNSTABLE](https://ocs4-jenkins-csb-ocsqe.apps.ocp4.prod.psi.redhat.com/job/qe-trigger-test-pr/642//testReport) (some or all tests failed).
TC passed, job is unstable due to tests/ecosystem/upgrade/test_upgrade.py::test_check_mon_pdb_post_upgrade which is part of the test upgrade validation and not relevant to this PR
image

**Comment by ramkiperiy** [comment]
@ebenahar @AviadP @petr-balogh Please review this TC and approve, this TC does scale of 1500 PVCs & PODs before upgrade of OCS and validates the scale post upgrade.
Please note TC will add 3 more worker nodes and these worker nodes are used for running app pods, OCS worker nodes are not used for running app pods.

**Review by Avilir — DISMISSED** [review]

**Comment by ramkiperiy** [comment]
@vasukulkarni @Avilir @AviadP @petr-balogh @ebenahar @keesturam Please review this TC and approve, I moved the TC to scale directory, earlier it was in ecosystem directory.
TC jenkins validation passed. 
Would like to know adding more worker nodes (for running app pods) does cause any issue to existing upgrade TC's ? these worker nodes were not labled with storage.

**Review by vasukulkarni — DISMISSED** [review]

File: tests/e2e/scale/upgrade/test_upgrade_with_scaled_pvcs_pods.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+# Each pod will be created with 20 PVCs attached
+PVCS_PER_POD = 20
+# Scale data file
+log_path = ocsci_log_path()
+SCALE_DATA_FILE = f"{log_path}/scale_data_file.yaml"
+
+
+@skipif_external_mode
+@skipif_bm
+@pre_ocs_upgrade
```
Comment (AviadP) [inline tests/e2e/scale/upgrade/test_upgrade_with_scaled_pvcs_pods.py:31]: im not sure this marker is correct - do we want to run this test for each upgrade job? in my opnion, new marker is needed here
@petr-balogh What do you think?

File: tests/e2e/scale/upgrade/test_upgrade_with_scaled_pvcs_pods.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        w_obj.write(str(f"POD_SCALE_LIST: {pod_running_list}\n"))
+        w_obj.write(str(f"PVC_SCALE_LIST: {pvc_bound_list}\n"))
+
+    # Check ceph health status
+    utils.ceph_health_check(tries=30)
+
+
+@skipif_bm
+@skipif_external_mode
+@post_ocs_upgrade
```
Comment (AviadP) [inline tests/e2e/scale/upgrade/test_upgrade_with_scaled_pvcs_pods.py:88]: same comment about this marker aswell

**Comment by ramkiperiy** [comment]
> PR validation
> 
> Cluster Name:
> Cluster Configuration:
> PR Test Suite: pre_upgrade or pre_ocs_upgrade or ocs_upgrade or post_ocs_upgrade or post_upgrade
> PR Test Path: tests/e2e/scale/upgrade/test_upgrade_with_scaled_pvcs_pods.py
> Additional Test Params:
> OCP VERSION: 4.8
> OCS VERSION: 4.7
> 
> [Job UNSTABLE](https://ocs4-jenkins-csb-ocsqe.apps.ocp4.prod.psi.redhat.com/job/qe-trigger-test-pr/1027//testReport) (some or all tests failed).

In above jenkins job scale tests passed without any issues, along with scale tests others were executed to see if there is any regression, 4 tests failed below is the analysis
1. tests.manage.z_cluster.cluster_expansion.test_add_capacity.TestAddCapacityPreUpgrade.test_add_capacity_pre_upgrade
    Failure: `E       AssertionError: Data re-balance failed to complete`
2. tests.manage.z_cluster.test_ceph_default_values_check.TestCephDefaultValuesCheck.test_check_mds_cache_memory_limit
    Failure: `E           ocs_ci.ocs.exceptions.CommandFailed: Error during execution of command: oc -n openshift-storage rsh rook-ceph-tools-98ffbd8c-njft2 ceph config show mds.ocs-storagecluster-cephfilesystem-a mds_cache_memory_limit --format json-pretty.
E           Error is Error ENOENT: no config state for daemon mds.ocs-storagecluster-cephfilesystem-a
E           command terminated with exit code 2`
3. tests.ecosystem.upgrade.test_resources.test_pod_io
    Failed: `E               ocs_ci.ocs.exceptions.TimeoutExpiredError: Timed out after 600s running get("pvc-test-bbfca8a84f984f09806e32ff5a87ba6", True, None)`
4. tests.manage.monitoring.test_workload_with_distruptions.test_workload_with_checksum_verify
    Failed: `E           subprocess.TimeoutExpired: Command '['oc', '-n', 'namespace-test-79aa790b88c54af98276c2f68', 'delete', 'Pod', 'delete-rbd-rwo-fs-2bcrl']' timed out after 600 seconds`

As you can note failures are not related to the scale test code, would suggest to merge this PR
image

**Comment by ramkiperiy** [comment]
@keesturam @vasukulkarni @petr-balogh @AviadP @ebenahar could you please review and approve this PR to merge.

File: ocs_ci/ocs/scale_lib.py [hunk 4]
```diff
@@ -1100,9 +1104,18 @@ def check_all_pod_reached_running_state_in_kube_job(
...
+                for i in pod_not_running_list:
+                    cmd = f"delete pod {i} -n {namespace}"
+                    ocp_obj.exec_oc_cmd(command=cmd, timeout=120)
+
+            # Breaking while loop after 13 Iteration i.e. after 30*13 secs of wait_time
             # And if PODs are still not in Running state then there will be assert.
-            if while_iteration_count >= 10:
+            if while_iteration_count >= 13:
                 assert logging.error(
                     f" Listed PODs took more than 300secs for Running {pod_not_running_list}"
```
Comment (keesturam) [inline ocs_ci/ocs/scale_lib.py:1120]: so the timeot would be 390 seconds now I suppose. can you update the error message accordingly.

**Review by Avilir — DISMISSED** [review]
LGTM

File: ocs_ci/ocs/scale_lib.py (hunk 4 above)
Comment (ramkiperiy) [reply ocs_ci/ocs/scale_lib.py]: done

**Review by Avilir — APPROVED** [review]

**Review by vasukulkarni — APPROVED** [review]

File: tests/e2e/scale/upgrade/test_upgrade_with_scaled_pvcs_pods.py (hunk 2 above)
Comment (ramkiperiy) [reply tests/e2e/scale/upgrade/test_upgrade_with_scaled_pvcs_pods.py]: Changed marker to pre_upgrade

File: tests/e2e/scale/upgrade/test_upgrade_with_scaled_pvcs_pods.py (hunk 3 above)
Comment (ramkiperiy) [reply tests/e2e/scale/upgrade/test_upgrade_with_scaled_pvcs_pods.py]: Changed marker to post_upgrade

**Review by ebenahar — APPROVED** [review]

**Review by keesturam — APPROVED** [review]

**Pull request merged**